File: apps/mobile_android/test/privacy_test.dart
import 'package:flutter_test/flutter_test.dart';
import '../lib/privacy.dart';

const _home = PrivacyZone(lat: 40.7128, lng: -74.006, radiusM: 200);

class _Pt {
  final double lat;
  final double lng;
  const _Pt(this.lat, this.lng);
}

double _lat(_Pt p) => p.lat;
double _lng(_Pt p) => p.lng;

/// Tiny offset that crosses outside a 200 m radius — about 350 m east.
_Pt _offset(double lat, double lng, double dLng) => _Pt(lat, lng + dLng);

void main() {
  group('isInAnyZone', () {
    test('empty zones returns false', () {
      expect(isInAnyZone(0, 0, const []), isFalse);
    });

    test('center is in zone', () {
      expect(isInAnyZone(_home.lat, _home.lng, const [_home]), isTrue);
    });

    test('far point is not', () {
      final p = _offset(_home.lat, _home.lng, 0.01);
      expect(isInAnyZone(p.lat, p.lng, const [_home]), isFalse);
    });
  });

  group('clipPointsToZones', () {
    test('empty zones returns input', () {
      final pts = const [_Pt(1, 1), _Pt(2, 2)];
      final out = clipPointsToZones<_Pt>(pts, const [],
          latOf: _lat, lngOf: _lng);
      expect(out, equals(pts));
    });

    test('drops leading + trailing in-zone', () {
      final pts = [
        _Pt(_home.lat, _home.lng), // in
        _Pt(_home.lat, _home.lng), // in
        _offset(_home.lat, _home.lng, 0.01), // out (mid)
        _offset(_home.lat, _home.lng, 0.02), // out (mid)
        _Pt(_home.lat, _home.lng), // in (trailing)
      ];
      final out = clipPointsToZones<_Pt>(pts, const [_home],
          latOf: _lat, lngOf: _lng);
      expect(out.length, 2);
      expect(identical(out[0], pts[2]), isTrue);
      expect(identical(out[1], pts[3]), isTrue);
    });

    test('keeps interior in-zone segments (only ends are clipped)', () {
      final pts = [
        _offset(_home.lat, _home.lng, 0.01), // out
        _Pt(_home.lat, _home.lng), // in (interior — kept)
        _offset(_home.lat, _home.lng, 0.02), // out
      ];
      final out = clipPointsToZones<_Pt>(pts, const [_home],
          latOf: _lat, lngOf: _lng);
      expect(out, equals(pts));
    });

    test('every point in zone returns empty', () {
      final pts = [
        _Pt(_home.lat, _home.lng),
        _Pt(_home.lat + 0.0001, _home.lng + 0.0001),
      ];
      final out = clipPointsToZones<_Pt>(pts, const [_home],
          latOf: _lat, lngOf: _lng);
      expect(out, isEmpty);
    });

    test('multiple zones — clips against the union', () {
      const work = PrivacyZone(lat: 40.75, lng: -73.99, radiusM: 200);
      final pts = [
        _Pt(_home.lat, _home.lng), // in home
        _offset(_home.lat, _home.lng, 0.01), // out (mid)
        _Pt(work.lat, work.lng), // in work (trailing)
      ];
      final out = clipPointsToZones<_Pt>(pts, const [_home, work],
          latOf: _lat, lngOf: _lng);
      expect(out.length, 1);
      expect(identical(out[0], pts[1]), isTrue);
    });
  });
}

File: apps/web/src/lib/privacy.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { clipPointsToZones, isInAnyZone, type PrivacyZone, type LatLng } from './privacy';

const home: PrivacyZone = { lat: 40.7128, lng: -74.006, radius_m: 200 };

// Tiny offset that crosses outside a 200m radius — about 350m east.
const offset = (lat: number, lng: number, dLng: number): LatLng => ({
	lat,
	lng: lng + dLng,
});

test('isInAnyZone — empty zones', () => {
	assert.equal(isInAnyZone({ lat: 0, lng: 0 }, []), false);
});

test('isInAnyZone — center is in zone', () => {
	assert.equal(isInAnyZone({ lat: home.lat, lng: home.lng }, [home]), true);
});

test('isInAnyZone — far point is not', () => {
	assert.equal(isInAnyZone(offset(home.lat, home.lng, 0.01), [home]), false);
});

test('clipPointsToZones — empty zones returns input', () => {
	const pts: LatLng[] = [{ lat: 1, lng: 1 }, { lat: 2, lng: 2 }];
	assert.deepEqual(clipPointsToZones(pts, []), pts);
});

test('clipPointsToZones — drops leading + trailing in-zone', () => {
	const pts: LatLng[] = [
		{ lat: home.lat, lng: home.lng }, // in
		{ lat: home.lat, lng: home.lng }, // in
		offset(home.lat, home.lng, 0.01), // out (mid)
		offset(home.lat, home.lng, 0.02), // out (mid)
		{ lat: home.lat, lng: home.lng }, // in (trailing)
	];
	const out = clipPointsToZones(pts, [home]);
	assert.equal(out.length, 2);
	assert.equal(out[0], pts[2]);
	assert.equal(out[1], pts[3]);
});

test('clipPointsToZones — keeps interior in-zone segments (only ends are clipped)', () => {
	const pts: LatLng[] = [
		offset(home.lat, home.lng, 0.01), // out
		{ lat: home.lat, lng: home.lng }, // in (interior — kept)
		offset(home.lat, home.lng, 0.02), // out
	];
	assert.deepEqual(clipPointsToZones(pts, [home]), pts);
});

test('clipPointsToZones — every point in zone returns empty', () => {
	const pts: LatLng[] = [
		{ lat: home.lat, lng: home.lng },
		{ lat: home.lat + 0.0001, lng: home.lng + 0.0001 },
	];
	assert.deepEqual(clipPointsToZones(pts, [home]), []);
});

test('clipPointsToZones — multiple zones', () => {
	const work: PrivacyZone = { lat: 40.75, lng: -73.99, radius_m: 200 };
	const pts: LatLng[] = [
		{ lat: home.lat, lng: home.lng }, // in home
		offset(home.lat, home.lng, 0.01), // out (mid)
		{ lat: work.lat, lng: work.lng }, // in work (trailing)
	];
	const out = clipPointsToZones(pts, [home, work]);
	assert.equal(out.length, 1);
	assert.equal(out[0], pts[1]);
});
